refactor: Adjust styling of back top top button

dsmmcken opened the pull request:
<!--
Thank you for sending the PR! We appreciate you spending the time to work on these changes.

Help us understand your motivation by explaining why you decided to make this change.

You can learn more about contributing to Docusaurus here: https://github.com/facebook/docusaurus/blob/main/CONTRIBUTING.md

Happy contributing!

-->

## Motivation

1. The new back to top button is in the same location as is common for live chat/help widgets, and there is no easy way to fix that without swizzling the "unsafe" component. This PR exposes position as css variables, so it's easy to shift the button to the side or above any other bottom right content. 
![image](https://user-images.githubusercontent.com/1576283/131858222-5be3c7cf-d689-4aef-83be-50a957b09599.png)

2. Subjective opinion: it is too visually prominent, I propose switching it from color-primary to color-secondary. The entry animation, plus the use of color-primary really draws your attention to it on entrance. It is a minor convenience feature, and should not act as a distraction from reading content.


Additionally: 
- it was missing an ARIA label, no indication of what this button was for a screen reader
- It lacked full themability: 
  - icon should be in css to allow customization, removed svg, can re-use an existing icon even. 
  - color shouldn't introduce a hardcode value, should use existing variable
  - buttons use color-dark for hover state rather than opacity, should match styling
  - should use box-shadow infima variable rather than custom shadow
  - posistion should be user themeable.


### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebook/docusaurus/blob/main/CONTRIBUTING.md#pull-requests)?

Yes.

## Test Plan

Button as now rendered in default theme
![image](https://user-images.githubusercontent.com/1576283/131859994-517a5e2f-02ce-46f6-bee1-05acd53b0b43.png)

I am unsure whether I did the change to base.json correctly. Couldn't figure out what the convention for naming them or if there was any logic to order, hope this was right. I just added it to the end.


## Related PRs

https://github.com/facebook/docusaurus/pull/5419

Josh-Cena commented:
Back to top button should actually be safe to swizzle, it's just because the whitelist hasn't been updated for long🤦‍♂️ cf #5263

slorber commented:
I'll be on holiday soon, but @lex111 can probably review this PR

Some quick thought
- we should mark it safe to swizzle
- we should add a stable className (`ThemeClassNames`) file so that it's easy to target with custom CSS
- adding new CSS vars would become useless
- didn't we remove this aria-label on purpose @lex111 ?

dsmmcken commented:
> * we should add a stable className (ThemeClassNames) file so that it's easy to target with custom CSS

That would be great, I haven't seen that adopted yet, what are the naming conventions?

> * didn't we remove this aria-label on purpose @lex111 ?

Title was removed, but it really should have an aria-label still for a11y, otherwise this button is a complete mystery as to what it does.

lex111 reviewed (commented):
Thanks for PR, but at the moment in custom CSS it is not possible to override CSS vars without using `!important` rule. Actually, I would prefer to use stable CSS class name instead of having bunch of CSS vars if users want to change default button style. `aria-label` is a really useful thing for a11y. Not sure about changing of background button , I'll have to think about that.

dsmmcken commented:
Oh interesting, I assumed because I can override --ifm variables, I could override --docusaurus variables. That seems like something I should be able to do. I see the inclusion order is different. Is there any ticket to improve this?

re: css vars vs stable class names. I would consider maintaining both. Having stuff pulled out as css variables make it easier for handling light/dark mode overrides, and it also kind of act as guide rails to users doing theming. It makes it clear, here are things you might need to customize, but if you need to do more than override the css directly.

Guidance for adding a stable class name? `docusaurus-button-back-to-top`?

lex111 commented:
Yes, we have dedicated issue to track this https://github.com/facebook/docusaurus/issues/3678

Most likely minority of users will customize back-to-top button, so I don't want to add even more CSS vars (we have so many  not very used ones already). So stable CSS class name seems like good compromise. This can be confusing, so we probably need to document more about CSS customization later.

For now suggest adding new class name here, such as `theme-back-to-top`. https://github.com/facebook/docusaurus/blob/main/packages/docusaurus-theme-common/src/utils/ThemeClassNames.ts#L35

About background, think you're right, at the moment the button is too noticeable. So we can use gray shade, by analogy with `button-secondary` button.

lex111 commented:
@dsmmcken please let me know if you cannot finish this PR, then I will do it myself.

dsmmcken commented:
@lex111 Was away for the long weekend, I can update it now that I am back.

dsmmcken commented:
Ready for re-review :
- dropped the root styles
- added a stable class name
- changed hover to use media queries detecting hover support, to address mobile issue from #5419. Previous fix would make hover state not apply on desktop after the first click, while still focused. https://css-tricks.com/solving-sticky-hover-states-with-media-hover-hover/
- added missing active state

lex111 commented on packages/docusaurus-theme-classic/src/theme/BackToTopButton/styles.module.css:
```diff
@@ -18,19 +13,33 @@
-  opacity: 0.8;
+.backToTopButton:after {
+  content: ' ';
+  mask: var(--ifm-menu-link-sublist-icon) 50% / 2rem 2rem no-repeat;
```
Why use mask and background-color, maybe just `background: var(--ifm-menu-link-sublist-icon) 50% 50% / 2rem 2rem no-repeat;` instead?

dsmmcken replied:
> Perhaps we need to add dark mode support also?

![image](https://user-images.githubusercontent.com/1576283/132520911-2c177461-88fa-4644-8e74-b38984d12a3d.png)

Better if I use the emphasis colours instead of secondary so it swaps colours?

Using mask allows you to directly re-colour the svg by just setting a background-color, and not having to edit underlying embedded svg or mess with filters like infima is doing for the sidebar arrows in light vs dark. The sidebar menu that uses this same icon should be made to do the same, but that code is on the infima side I think.

lex111 replied:
Interesting idea with using mask, although it requires vendor prefix to work. Maybe it makes sense to use mask instead of filter in Infima..

Agree about using emphasis colors, but in dark mode the button background is pretty bright, maybe use 200 instead of 300? Then the button would look like collapse sidebar button by color. WDYT?

| Before   | After    |
| -------- | -------- |
| ![image](https://user-images.githubusercontent.com/4408379/132530736-5a9bc3d3-530d-41f1-8d52-8fb384acf75b.png) | ![image](https://user-images.githubusercontent.com/4408379/132531546-9aaff6d1-6778-4641-9ec0-ddd1fe08a68f.png) |

dsmmcken replied:
![image](https://user-images.githubusercontent.com/1576283/132576949-47cdde12-8fd5-442b-ba35-eecb9816ccc2.png)

Stepped down to 200.  Cool now?

The collapse bar is not part of the color scale? I've never noticed, I have it turned off.

lex111 replied:
Yes, it looks great now.

Collapse button uses the built-in button classes from Infima, `button button-secondary', to which dark mode support has been added in Docusaurus code. Perhaps we should do the same with back-to-top button?

dsmmcken replied:
Sorry, is there something else you want changed in this PR or is it good to go?

lex111 replied:
Yes, your PR is fine, just added the new message to translate in supported languages and removed unnecessary CSS code.

lex111 replied:
I also removed `active` state you added, because for now it actually does not work -- we need to add `!important` rule for this, but I think this state is not too important for such small button.

lex111 reviewed (commented):
Perhaps we need to add dark mode support also? 

Current state:

![image](https://user-images.githubusercontent.com/4408379/132504969-857f07f6-ebc0-47b8-988c-d3c1cfd25d18.png)

lex111 commented:
@dsmmcken thanks!

slorber commented on packages/docusaurus-theme-classic/codeTranslations/base.json:
```diff
@@ -96,7 +97,7 @@
   "theme.docs.versions.latestVersionLinkLabel___DESCRIPTION": "The label used for the latest version suggestion link label",
   "theme.docs.versions.latestVersionSuggestionLabel": "For up-to-date documentation, see the {latestVersionLink} ({versionLabel}).",
-  "theme.docs.versions.latestVersionSuggestionLabel___DESCRIPTION": "The label used to tell the user to check the latest version",
+  "theme.docs.versions.latestVersionSuggestionLabel___DESCRIPTION": "The label userd to tell the user that he's browsing an unmaintained doc version",
```
FYI this description change has a typo and will be overridden anyway, as it's extracted from the code
